src/components/withNetwork.js

```diff
@@ -0,0 +1,76 @@
+import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+import {withOnyx} from 'react-native-onyx';
+import getComponentDisplayName from '../libs/getComponentDisplayName';
+import ONYXKEYS from '../ONYXKEYS';
+import networkPropTypes from './networkPropTypes';
+
+/**
+ * An HOC that passes current network status as a prop to the WrappedComponent
+ *
+ * When transitioning from offline -> online, this.props.network.isReconnecting
+ * will be true.
+ *
+ * @param {Component} [WrappedComponent] the component that needs network status
+ * @returns {Component} the component with network status provided
+ */
+export default function (WrappedComponent) {
+    const propTypes = {
+        forwardedRef: PropTypes.func,
+
+        /** Information about the network */
+        network: networkPropTypes.isRequired,
+    };
+
+    const defaultProps = {
+        forwardedRef: () => {},
+    };
+
+    class WithNetwork extends Component {
+        constructor(props) {
+            super(props);
+            this.isReconnecting = false;
+        }
+
+        shouldComponentUpdate(nextProps) {
+            // we check for reconnecting here so that we can pass this as a prop on render
+            this.isReconnecting = this.props.network.isOffline && !nextProps.network.isOffline;
```

Comment:
Every component using this HOC is still going to have to do:

```
componentDidUpdate(prevProps) {
    if (prevProps.isReconnecting && !this.props.isReconnecting) {
        // fetch data
    }
}
```

So this HOC really isn't providing any value.

Reply:
The way I had envisioned this HOC would work is like this:
```
componentDidUpdate(prevProps) {
  if (prevProps.network.isOffline && !this.props.network.isOffline) {
    if (_.isFunction(this.props.forwardedRef.onReconnect)) {
        this.props.forwardedRef.onReconnect();
    }
  }
}
```

My BIG DISCLAIMER here is that I am such an amateur when it comes to the forward refs that they are hard for me to wrap my brain around. But the idea is there. You need to access the `ref` of the wrapped component, and call the `onReconnect()` method.

Reply:
Maybe @marcaaron can figure out what I'm trying to get at 😬

Reply:
There are some options. Forwarding refs didn't come to mind, but could work.

I think the idea should be to encapsulate this behavior:

>When the network state changes from offline to online trigger some arbitrary code defined by another component

First couple of solutions that come to mind...

**Use an HOC but register a reconnect callback in the constructor**

```js
class MyComponent extends Component {
    constructor(props) {
        super(props);

        // where onReconnect is passed as a prop via `withReconnect()`
        props.onReconnect(this.doSomeCustomLogic.bind(this));
    }
    ...
}
export default withReconnect(MyComponent);
```

IMO, doesn't work so good for non class components and is pretty verbose

**Use a component instead of an HOC**

```js
<Reconnect onReconnect={this.doSomeCustomLogic} />
```

I like this much better as it is more react-y in that we are getting a child component to handle some responsibility instead of forcing a parent to run logic defined in a child. And I think it's easier to reason about even though the component does not render any views.

Reply:
Oh, that last idea is kind of cool. I'm always afraid to propose components that don't render anything, but it is definitely elegant.

Reply:
I'm pretty open to the component idea since React components are "just functions". At least I can't think of why it would be bad to not render anything (maybe slightly confusing because you expect a component to render something). 

If we don't like that... another completely different way to handle this would be to use the `NetworkConnection` [event that detects offline -> online and triggers reconnection callbacks](https://github.com/Expensify/App/blob/ba5a38f4a8f8eca2fdc63eb2b45aaccc85fe7997/src/libs/NetworkConnection.js#L39-L43) to also run some callbacks.

Possibly something like this...
 
```js
const componentCallbacks = {};
let count = 0;
function onComponentReconnect(callback) {
    const id = count;
    count++;
    componentCallbacks[id] = callback;
     return () => {
       delete componentCallbacks[id];
    };
}
```

Then we could use these in `componentDidMount()` and `componentWillUnmount()` like this:

```js
componentDidMount() {
    this.unsubscribe = NetworkConnection.onComponentReconnect(this.doSomething);
}

componentWillUnmount() {
    this.unsubscribe();
}
```

And then something like 

```js
    // When reconnecting, ie, going from offline to online, all the reconnection callbacks
    // are triggered (this is usually Actions that need to re-download data from the server)
    if (isOffline && !isCurrentlyOffline) {
        triggerReconnectionCallbacks('offline status changed');
        _.each(componentCallbacks, callback => callback());
    }
```